### ISSUE TITLE
Add "Language Reference"

### ISSUE THREAD

**Issue by LandonSchropp** [google/jsonnet#587]
On the [Language Reference](https://jsonnet.org/ref/language.html) section of the Jsonnet website, it says this:

> TODO: This page still needs to be written.

Jsonnet would be much easier to use with a language reference/API documentation. I was surprised that it wasn't there, and not including it has made some things that should have been simple difficult to figure out.

**Comment by sbarzowski** [comment]
Thanks for the feedback. Just to clarify most of things mentioned there are covered by the formal spec here (https://jsonnet.org/ref/spec.html), though it's not very easy to read, especially if you're not used to the notation. 

Can you tell us what specifically was hard to figure out for you? We'll need to prioritize, so it's good to know what's most problematic.

**Comment by sparkprime** [comment]
Yes please tell us what you expected to find so we can make sure it's included.

Specifically what things that you found difficult to figure out.

**Comment by LandonSchropp** [comment]
Sorry for the slow reply! I just got back from vacation.

For the first point, I definitely couldn't wrap my head around the notation in the formal specification. The specific issue I was trying to solve when I was digging through the documentation was whether I could export a function from a file directly without first having to store it in a variable. I searched through the spec, but I didn't have the first idea of how to find what I was looking for in there.

What I was hoping for something like a [Learn X in Y Minutes](https://learnxinyminutes.com/docs/ruby/) guide for Jsonnet. It would be great to have a simple document that outlines the language syntax at a high-level. The tutorial does this to a degree, but it felt too specific. I could answer some questions by reading it, but not others.

**Comment by nrclark** [comment]
+1 agree that I'd love to see the language reference get written!

I'm writing a tool that uses JSON mapfiles for configuration, and the mapfiles can get pretty complex. Jsonnet looks like the perfect tool - I think? Except that maybe the documentation's not so clear on how to use it. How does a `for`-loop work? What comes from `std` and what's part of the Jsonnet language? How do I write a multi-line string (if it's possible)?

I can puzzle through some things from the `Tutorial` and `Standard Library` sections. As a user, how do I know what I don't know, though? Is all of Jsonnet's functionality described in those two documents? And if I want to promote the tool for use on my team, can I ask them in fairness to puzzle through what is/isn't possible based on the tutorial?

It seems like the `Language Reference` page on `jsonnet.org` maybe has been stale for a while now. If a reference isn't on anybody's agenda, should be page be removed?

**Comment by sbarzowski** [comment]
Thanks for the feedback.

> I'm writing a tool that uses JSON mapfiles for configuration, and the mapfiles can get pretty complex. Jsonnet looks like the perfect authoring tool - I think?

I'm not sure what exactly you mean by mapfiles, but generating complex JSON (or other structured formats, especially hierarchical ones) sounds well suited for Jsonnet. It is usually used to write some glue between the target format and how you would like to describe your configuration / whatever.

> Except that maybe the documentation's not so clear on how to use it. How does a for-loop work? 

I'm going to answer these questions here. I know you made a general point, but maybe it'll be helpful :-). 

There is an array and object comprehension syntax, which is described in https://jsonnet.org/learning/tutorial.html#comprehension. This is the only "for-loop" possible in Jsonnet.

> What comes from std and what's part of the Jsonnet language?

Not sure I understand the question. Stuff described in https://jsonnet.org/ref/stdlib.html is the stdlib. In the code you can always tell, because the standard library is exposed as `std` object, so all stdlib calls have the `std.something` form.

> How do I write a multi-line string (if it's possible)?

Like that:
```
local myVar = |||
  foo
  bar
|||
```
Tutorial covers it here: https://jsonnet.org/learning/tutorial.html#syntax.

> I can puzzle through some things from the Tutorial and Standard Library sections. As a user, how do I know what I don't know, though? Is all of Jsonnet's functionality described in those two documents? 

There is a really precise documentation in the Jsonnet Specification section. If you want to check what is possible at all, this is your thing. Actually just looking at the initial part (the grammar) is going to tell you what programs are possible. The problem is that unfortunately not everyone is familiar with the formalism used there. 

I think the reference was intended to be something in the middle between the tutorial and the spec. There is also a matter of explaining patterns and guidance on how to use the concepts that Jsonnet provides. Jsonnet is a *very* simple language and its power comes from composing simple concepts. It may not be immediately obvious how to achieve something in Jsonnet, especially without any prior experience with functional languages. 

**Also, if something is not clear, asking questions on Stackoverflow tends to work well (most questions get a comprehensive answer in <24h).** If something looks particularly tricky or you suspect it's broken, you can also open an issue here.

> And if I want to promote the tool for use on my team, can I ask them in fairness to puzzle through what is/isn't possible based on the tutorial?

I think this is actually pretty fair. Tutorial + Stdlib covers 90% of the functionality. There is also a spec if you want to figure out some boundary case. And you can always ask questions.

> It seems like the Language Reference page on jsonnet.org maybe has been stale for a while now, and potentially reduces adoption because of the optics. If a reference isn't on anybody's agenda, should be page be removed?

Yes, probably. Instead of removing, it may be incorporated back into the spec section.

Anyway, I'll try adding a few sections over the next week or two. We'll see how it goes and then we will decide. **Any volunteers for proofreading?**

**Comment by nrclark** [comment]
Sure, I'll give a read! Plus I can offer the perspective of a greenhorn.

**Comment by dsabsay** [comment]
I'll proofread as well. Just tag me in a PR and I'll gladly take a look.

**Comment by johnbartholomew** [comment]
Closing out old issues. I think this one can be considered "done" to a reasonable extent. The documentation on jsonnet.org at this time includes:
- A tutorial (including interactive pieces)
- A language guide
- The formal spec
- Standard library reference documentation.
